arch/x86/pagetables.c

```diff
@@ -420,6 +421,26 @@ void *vmap_4k(cr3_t *cr3_ptr, void *va, mfn_t mfn, unsigned long l1_flags,
     return va;
 }
 
+void *vmap_frame_refill(void *va, mfn_t mfn, bool paging_lock) {
+    unsigned long _va = _ul(va) & PAGE_ORDER_TO_MASK(PAGE_ORDER_4K);
+
+    dprintk("%s: va: 0x%p mfn: 0x%lx\n", __func__, va, mfn);
+
+    if (paging_lock)
+        spin_lock(&vmap_lock);
+    else
+        ASSERT(vmap_lock);
+
+    va = _vmap_4k(&cr3, _ptr(_va), mfn, L1_PROT, false);
+
+    if (paging_lock)
+        spin_unlock(&vmap_lock);
+    else
+        ASSERT(vmap_lock);
+
+    return va;
+}
+
```

Comment:
nit: instead of the `paging_lock` check I would create two versions of this function: one with locking and another with the ASSERT (or maybe a BUG_ON). Despite code duplication, it probably will pay off later when debugging something else.